### PR TITLE
feat(curriculum): Allow use of triple backticks for code in description and instructions sections of challenges

### LINKED ISSUE #35913: Recommend using semantic markup for code
In the description and instruction sections of the challenges, the style guide recommends `<blockquote>` to markup blocks of code, and `<code>` for inline code.  

I think it would be better to be consistent and use semantic markup for both.

We could either do this by through `<pre><code>` tags or take advantage of the fact we're using markdown and use code fences, i.e. \```js ```.  This would have the benefit that we could use prismjs to make the code highlighting in the challenge descriptions the same as in the guide.

### PR DESCRIPTION
**Pull request by RandellDawson**
- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] None of my changes are plagiarized from another source without proper attribution.
- [x] All the files I changed are in the same world language (for example: only English changes, or only Chinese changes, etc.)
- [x] My changes do not use shortened URLs or affiliate links.

Closes #35913

This PR is an attempt to allow markdown code sections in the `Description` and  `Instructions` sections of the curriculum challenge files.  Currently, you must use `&nbsp;` and `<br>` tags inside `blockquote` to make the code sections look indented and appear on new lines.   With the changes made the following CSS files, I was able to almost duplicate the look and feel of the current blockquote CSS.  I say "almost", because in Nite-mode, I can not figure out how to get the left white line on blockquote elements in the description and instructions sections.  Honestly, I am not even sure if the left white line should even show in the current blockquote as it does locally and in current production.

* client/src/components/layouts/night.css
* client/src/templates/Challenges/components/challenge-description.css

For this first attempt, I only converted the 18 challenges in the `Basic Data Structures` section which had `blockquote` elements in the description and instructions sections.  I used a custom script locally to make the conversions.  If we agree this PR looks good and is merge-able, then I can either run the script on the other challenges or create separate PRs for each section, to keep the number of files changed smaller and make them easier to review.

### CODE REVIEW

**Comment by ojeytonwilliams** [comment]
I don't particularly like the white line, either.  If we want it gone, it's a border, so we could use
```css
.night .challenge-instructions pre {
  border-left: 5px solid #222;
}
```
to make it look like part of the background.  Or just remove it from both night and day modes and adjust the padding.

**Comment by ojeytonwilliams** [comment]
@RandellDawson This should make all three approaches - blockquotes, backticks and backticks with a specific language - look the same.

Please don't clean up the commit history just yet, I need to figure out what I keep doing wrong!

**Comment by RandellDawson** [comment]
@raisedadead, @ValeraS Hoping you could lend your opinion on this PR.  Thanks.

**Comment by RandellDawson** [comment]
@ojeytonwilliams I fixed a merge conflict and then added `js` to the end of the same challenges where I replaced the `blockquote` with 3 backticks.  I noticed there is something still not correct.  On challenge `basic-data-structures/add-items-to-an-array-with-push-and-unshift`, there is a horizontal scrollbar showing which should not be there.

image

Make sure to pull down my latest changes and see if you can fix the CSS.

**Comment by ojeytonwilliams** [comment]
@RandellDawson Well spotted!  This should be fixed, along with the slightly too large lines.  Let me know if you catch anything else, though hopefully we've found all the discrepancies now.

**Comment by RandellDawson** [comment]
@ojeytonwilliams I think your last changes take care of all the formatting issues I could find.  

If we could have @Manish-Giri, @moT01, and @thecodingaviator  take a look and offer their thoughts that would be great.

**Review by Manish-Giri — COMMENTED** [review]
I went through all of the changed challenges in the Basic Data Structures section on Chrome and Safari, and everything looks great!

I wasn't able to test the Night Mode yet (due to sign-in issues), planning to do that later today.

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-data-structures/add-key-value-pairs-to-javascript-objects.english.md
Comment (moT01) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-data-structures/add-key-value-pairs-to-javascript-objects.english.md]:
```suggestion
let userData = FCC_User['followers'];
```

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-data-structures/check-for-the-presence-of-an-element-with-indexof.english.md
Comment (moT01) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-data-structures/check-for-the-presence-of-an-element-with-indexof.english.md]:
```suggestion
fruits.indexOf('dates'); // returns -1
```

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-data-structures/check-for-the-presence-of-an-element-with-indexof.english.md
Comment (moT01) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-data-structures/check-for-the-presence-of-an-element-with-indexof.english.md]:
```suggestion
fruits.indexOf('oranges'); // returns 2
```

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-data-structures/check-for-the-presence-of-an-element-with-indexof.english.md
Comment (moT01) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-data-structures/check-for-the-presence-of-an-element-with-indexof.english.md]:
```suggestion
fruits.indexOf('pears'); // returns 1, the first index at which the element exists
```

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-data-structures/copy-array-items-using-slice.english.md
Comment (moT01) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-data-structures/copy-array-items-using-slice.english.md]:
```suggestion
```

**Review by moT01 — DISMISSED** [review]
@RandellDawson I put some things in there that could probably be fixed quick. I know the missing semi-colons were already missing and not from you, and the extra line on the one doesn't actually affect the looks. So fix them if you want, or they could just be left alone.

Overall the formatting looks good. Everything seems to line up and all.
I think the red text color should be green on night mode. I'm not sure if that was your call or maybe beta is already doing that (I can't check there right now)

**This PR:**
Screen Shot 2019-05-01 at 9 07 01 PM

**Current Production:**
Screen Shot 2019-05-01 at 9 17 39 PM

The red really probably looks fine, I think it should be the dark green though.

**Comment by RandellDawson** [comment]
@moT01 My goal was not to map out the best nite mode look.  I strictly was trying to emulate what the current day and nite modes look like with the current use of `blockquote`.  I think deciding the colors should be discussed in a separate issue and then we can always update the nite mode color CSS (that is extremely easy).

Also, I don't mind adding the semi-colons (I committed your suggestions), but the purpose of this PR was to convert the `blockquote` elements in the Description and Instructions section "as-is".  If I add the conversion of the other 151 challenges which currently have `blockquote` elements in the `JavaScript Algorithms and Data Structures` section on this PR, I will do my best to look for missing semi-colons, but I think we should keep the scope focused on the conversion and not the existing code. 

I was hoping the review would stick to validating the conversion was correct and not other things that might be missing from the original code being converted.

If others disagree with my approach above and think we should fix things beyond the conversion part, then I can do it, but it could take a while to get this PR merged.  I really like to keep a PRs review to a specific scope, otherwise the comments/change requests can become overwhelming and difficult to track on the PR itself.  

These changes requests were just for converting 18 challenges, so it is quite possible converting the remaining 151 of this section, would have many more such change requests for other extra lines and missing semi-colons.  If it turns out that there are only 15-20 "fixes" after converting the remaining 151, then I would have no issue taking care of them, but my guess is there will be many more.

**Comment by ojeytonwilliams** [comment]
I just want to go over some of what @RandellDawson and I have discussed, so we're all on the same page.  

The original idea was that this PR would function as a bridge from blockquotes to highlighted code.  The first step was to change the css so we can create challenges with backticks rather than having to use blockquotes.  That's pretty much done.  Once that's merged we can change the existing challenges to use backticks, which is what @RandellDawson has been focussing on.  Some of that's been folded into this PR, so we can check everything looks good.  

Once all the challenges have been updated, we can consider highlighting them with Prism, like the guide is currently.  This dovetails nicely with the command line chic redesign.  To that end, the plan was to add the code language (i.e ```js), so we are able to highlight them later, but leave the visuals unchanged for now.

**Comment by thecodingaviator** [comment]
@ojeytonwilliams @RandellDawson I think the coloring should be left up to @ahmadabdolsaheb and both of you could check with each other on the colors and how Ahmad is doing it on the new night mode theme.

**Comment by moT01** [comment]
I believe the current beta has red text for the blockquote code blocks on night mode. I'm not a fan, but that's how it already was. However it does look like this PR changes the inline code blocks from green to red - so I needed to bring that up.

**Review by thecodingaviator — COMMENTED** [review]
I checked how the thing looks on the normal mode, looks good to me, as for the dark mode, I haven't checked. What I think is that we should leave the color styling up to the changes that are being made in the other PR by @ahmadabdolsaheb

File: client/src/templates/Challenges/components/challenge-description.css
Comment (ValeraS) [inline client/src/templates/Challenges/components/challenge-description.css]: @ojeytonwilliams Why do we need the last selector?

Comment (ValeraS) [reply client/src/templates/Challenges/components/challenge-description.css]: Is `.challenge-instructions pre` not enough?

Comment (RandellDawson) [reply client/src/templates/Challenges/components/challenge-description.css]: @ValeraS We tried that originally, but it did not work.

Comment (ValeraS) [reply client/src/templates/Challenges/components/challenge-description.css]: Prism has these selectors: `code[class*="language-"], pre[class*="language-"]`
`.challenge-instructions pre` selector weights more.

Triple backticks with lang become `<pre><code class="language-lang">...`

What construction becomes `<pre class ="language-...">`?

Comment (RandellDawson) [reply client/src/templates/Challenges/components/challenge-description.css]: @ojeytonwilliams As a test, I removed just `.challenge-instructions pre[class*="language-"] ` as a selector and left all other CSS the same in the `challenge-descriptions.css` file.  It seems to be fine.  Can you validate also?  I am not sure why I thought it did not work originally, but maybe it had to do with other CSS we had not added at that point.

Comment (ojeytonwilliams) [reply client/src/templates/Challenges/components/challenge-description.css]: > What construction becomes `<pre class ="language-...">`?

@ValeraS  Good point - none of the constructions we're trying to support.  You would have to put pre tags in manually, rather than than via backticks, so we don't need this.

@RandellDawson Exactly.  We needed some more css, just not this bit.

**Comment by ojeytonwilliams** [comment]
@moT01 @ValeraS This last commit should fix the problems you pointed out.  In the process of fixing them, I noticed some other minor issues (the font and line height) and sorted those out, too.

**Review by ojeytonwilliams — APPROVED** [review]

File: client/src/templates/Challenges/components/challenge-description.css
Comment (ValeraS) [inline client/src/templates/Challenges/components/challenge-description.css]:
```suggestion
  font-family: 'SFMono-Regular', Consolas, 'Roboto Mono', 'Droid Sans Mono', 'Liberation Mono', Menlo, Courier, monospace;
```

File: client/src/templates/Challenges/components/challenge-description.css
Comment (ValeraS) [inline client/src/templates/Challenges/components/challenge-description.css]:
```suggestion
  text-shadow: none;
```

File: client/src/templates/Challenges/components/challenge-description.css
Comment (ValeraS) [inline client/src/templates/Challenges/components/challenge-description.css]:
```suggestion
  line-height: 1.42;
```

File: client/src/templates/Challenges/components/challenge-description.css
Comment (ValeraS) [inline client/src/templates/Challenges/components/challenge-description.css]:
```suggestion
  border: none;
```

File: client/src/templates/Challenges/components/challenge-description.css
Comment (ValeraS) [inline client/src/templates/Challenges/components/challenge-description.css]: IMO, It's more common, do not break code lines.

File: client/src/templates/Challenges/components/challenge-description.css
Comment (ValeraS) [inline client/src/templates/Challenges/components/challenge-description.css]:
```suggestion
.challenge-instructions blockquote,
.challenge-instructions pre {
```

File: client/src/components/layouts/night.css
Comment (ValeraS) [inline client/src/components/layouts/night.css]: unite with `.night code` and use its styles.

**Review by ValeraS — CHANGES_REQUESTED** [review]

File: client/src/templates/Challenges/components/challenge-description.css
Comment (RandellDawson) [reply client/src/templates/Challenges/components/challenge-description.css]: So what is your actual code change suggestion for this comment?

Comment (ValeraS) [reply client/src/templates/Challenges/components/challenge-description.css]: do not change `white-space`, `word-break`, `word-wrap`

Comment (RandellDawson) [reply client/src/templates/Challenges/components/challenge-description.css]: But then it will not function the same as the current `blockquote` which is the whole point of this PR.

Comment (ValeraS) [reply client/src/templates/Challenges/components/challenge-description.css]: If you decide that code blocks should simulate a blockquote, then OK, otherwise why add it here and then delete it in another PR?

**Comment by ojeytonwilliams** [comment]
This new commit includes the changes @ValeraS requested.  The code blocks now look more conventional in that they don't allow word wrapping.  

@raisedadead Could you take a quick look, as the challenges will look slightly different?

**Comment by raisedadead** [comment]
@ojeytonwilliams thanks for bringing this to my attention.

This PR does change a lot of files so it will be a while before I can review all the code, and I do not want to hold you all up due to my ongoing tasks. Can you please post some screenshots to help me understand how this looks?

We can then merge and deploy this to beta for everyone to test, leave feedback and make any additional changes.

**Comment by ojeytonwilliams** [comment]
@raisedadead Sure, I understand.

All the following are from learn/javascript-algorithms-and-data-structures/basic-data-structures/check-for-the-presence-of-an-element-with-indexof

This is how a code block, ```js, looks:
CodeDay
Night mode:
CodeNight
This is the original page, using blockquotes, but with the new css:
BlockDay
and again in night mode:
BlockNight

We weren't sure about leaving blockquotes red, but we decided to make as few changes as possible.

**Comment by raisedadead** [comment]
I think the red in the blockquote in night mode will not pass the AAA rating that we strive for a11y, Can we possibly use the green as seen in the  inline code tags (only in the night node) ?

**Comment by raisedadead** [comment]
Other than that I am good to merge this and deploy for feedback.

**Comment by ojeytonwilliams** [comment]
I went ahead and grouped up the various types of blocks.  I ran them through a contrast analyser and they manage 4.5:1 contrast, i.e. AAA level, assuming the font is large, but not if it's small.  The green is too dark, even for a black background.  I hope that's okay, since it's easy enough to get your browser to expand the text.

**Review by raisedadead — APPROVED** [review]

**Comment by raisedadead** [comment]
Thanks for making these changes, I am going to pull this down locally, make a quick sanity check and prep a build. I should have this deployed in the next few hours, if all looks good.

Thanks for everyone pitching in on this.

**Comment by RandellDawson** [comment]
@raisedadead It has been a team effort for sure.

**Comment by RandellDawson** [comment]
@raisedadead Can I go ahead and create PRs for the various sections which converts the blockquuotes to the triple backticks?  I already created a script to do so.

**Comment by raisedadead** [comment]
You should wait for the deployment to complete and test if things look good.

**Comment by raisedadead** [comment]
Deployed on beta site.

**Comment by RandellDawson** [comment]
@ojeytonwilliams I just checked the beta site and I thought we had it so there was no horizontal scroll bar on the code sections.  It should wrap like the blockquotes do.

**Comment by ojeytonwilliams** [comment]
We did, initially, but the blockquotes and code diverged in that respect during the review back and forth. Do they look okay?

**Comment by RandellDawson** [comment]
I think it looks better if the code wraps which the blockquote does do.  Other than that, I think everything looks fine.

**Pull request merged**
